libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+    
+    if (!MCStringFormat(&t_path, "%s/jre/lib/jli/libjli.dylib", t_javahome))
+        return false;
+#endif
+    
+    MCAutoStringRefAsSysString t_jvm_lib;
+    if (!t_jvm_lib . Lock(*t_path))
+        return false;
+    
+    if (!initialise_weak_link_jvm_with_path(*t_jvm_lib))
+        return false;
+    
+    s_weak_link_jvm = true;
+    return true;
+}
+
+#endif
+
+static void init_jvm_args(JavaVMInitArgs *x_args)
+{
+#if defined(TARGET_PLATFORM_LINUX) || defined(TARGET_PLATFORM_MACOS_X)
+    JNI_GetDefaultJavaVMInitArgs(x_args);
+#endif
+}
+
+static bool create_jvm(JavaVMInitArgs *p_args)
+{
+    int ret = 0;
+#if defined(TARGET_PLATFORM_LINUX) || defined(TARGET_PLATFORM_MACOS_X)
+    ret = JNI_CreateJavaVM(&s_jvm, (void **)&s_env, p_args);
+#endif
+    
+    return ret == 0;
+}
+
+bool initialise_jvm()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    if (!initialise_weak_link_jvm())
+        return false;
+    JavaVMInitArgs vm_args;
+    vm_args.version = JNI_VERSION_1_6;
+    init_jvm_args(&vm_args);
+    
+    JavaVMOption* options = new (nothrow) JavaVMOption[1];
+    options[0].optionString = "-Djava.class.path=/usr/lib/java";
+    
+    vm_args.nOptions = 1;
+    vm_args.options = options;
+    vm_args.ignoreUnrecognized = false;
+    
+    return create_jvm(&vm_args);
+#endif
+    return true;
+}
+
+void finalise_jvm()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    if (s_jvm != nullptr)
+    {
+        s_jvm -> DestroyJavaVM();
+    }
+#endif
+}
+
+void MCJavaDoAttachCurrentThread()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    s_jvm -> AttachCurrentThread((void **)&s_env, nullptr);
+#else
+    s_env = MCJavaAttachCurrentThread();
+#endif
+}
+
+static bool __MCJavaStringFromJString(jstring p_string, MCStringRef& r_string)
+{
+    MCJavaDoAttachCurrentThread();
+    const char *nativeString = s_env -> GetStringUTFChars(p_string, 0);
+    bool t_success = MCStringCreateWithCString(nativeString, r_string);
+    s_env->ReleaseStringUTFChars(p_string, nativeString);
+    return t_success;
+}
+
+static bool __MCJavaStringToJString(MCStringRef p_string, jstring& r_string)
+{
+    MCJavaDoAttachCurrentThread();
+    
+    MCAutoStringRefAsCString t_string_cstring;
+    t_string_cstring . Lock(p_string);
+    
+    jstring t_result;
+    t_result = s_env->NewStringUTF(*t_string_cstring);
+    
+    if (t_result != nullptr)
+    {
+        r_string = t_result;
+        return true;
+    }
+    
+    return false;
+}
+
+static bool __MCJavaDataFromJByteArray(jbyteArray p_byte_array, MCDataRef& r_data)
+{
+    MCJavaDoAttachCurrentThread();
+    
+    jbyte *t_bytes = nullptr;
+    uint32_t t_length = 0;
+    
+    if (p_byte_array != nullptr)
+        t_bytes = s_env -> GetByteArrayElements(p_byte_array, nullptr);
+    
+    if (t_bytes == nullptr)
+        return false;
+
+    t_length = s_env -> GetArrayLength(p_byte_array);
```

Comment:
`t_length` should be a `jsize`, and I think you need to check for overflow when converting to a `uindex_t` for the call to `MCDataCreateWithBytes` because `jsize` is 64 bits long on 64-bit platforms.  Might be a good idea to do a quick audit of JNI methods that return `jsize` to check for possible overflows.